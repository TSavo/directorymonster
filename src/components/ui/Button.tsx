'use client';

import React from 'react';
import { VariantProps } from 'class-variance-authority';
import { useButton } from './hooks/useButton';
import ButtonPresentation from './ButtonPresentation';

// Re-export the button variants from the hook
export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<any> {
  isLoading?: boolean;
  asChild?: boolean;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  loadingText?: string;
}

/**
 * Button Component
 *
 * A reusable button component with different variants and sizes
 */
export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (props, ref) => {
    const {
      shouldRenderAsChild,
      asChildProps,
      buttonProps,
      showSpinner,
      showLeftIcon,
      showRightIcon,
      buttonText,
      leftIcon,
      rightIcon
    } = useButton(props, ref);

    // If asChild is true, we render the children directly with the button props
    if (shouldRenderAsChild && React.isValidElement(props.children)) {
      return React.cloneElement(props.children, asChildProps);
    }

    // Otherwise, render the button presentation component
    return (
      <ButtonPresentation
        buttonProps={buttonProps}
        showSpinner={showSpinner}
        showLeftIcon={showLeftIcon}
        showRightIcon={showRightIcon}
        buttonText={buttonText}
        leftIcon={leftIcon}
        rightIcon={rightIcon}
        ref={ref}
      />
    );
  }
);

Button.displayName = "Button";

export default Button;
