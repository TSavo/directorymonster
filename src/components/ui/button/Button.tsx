'use client';

import * as React from 'react';

export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: 'default' | 'destructive' | 'outline' | 'secondary' | 'ghost' | 'link';
  size?: 'default' | 'sm' | 'lg' | 'icon';
  asChild?: boolean;
  isLoading?: boolean;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  loadingText?: string;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ 
    className, 
    variant = 'default', 
    size = 'default', 
    asChild = false, 
    isLoading = false, 
    leftIcon, 
    rightIcon, 
    loadingText, 
    children, 
    ...props 
  }, ref) => {
    const Comp = asChild ? 'div' : 'button';
    return (
      <Comp
        className={className}
        ref={ref}
        data-variant={variant}
        data-size={size}
        data-loading={isLoading || undefined}
        {...props}
      >
        {leftIcon && <span data-testid="left-icon">{leftIcon}</span>}
        {isLoading && loadingText ? loadingText : children}
        {rightIcon && <span data-testid="right-icon">{rightIcon}</span>}
      </Comp>
    );
  }
);
Button.displayName = 'Button';
