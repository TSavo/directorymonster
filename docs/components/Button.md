# Button Component

The Button component is a versatile, reusable UI element that provides a consistent look and feel across the application. It supports various variants, sizes, and states to accommodate different use cases.

## Features

- **Multiple Variants**: Primary, Secondary, Danger, Ghost, and Link
- **Multiple Sizes**: Small, Medium, Large, and Icon
- **Loading State**: Visual indicator for asynchronous operations
- **Composition**: Can wrap other components like `Link` using the `asChild` prop
- **Accessibility**: Proper focus states and ARIA attributes

## Usage

```tsx
import { Button } from '@/components/ui/Button';

// Basic usage
<Button>Click Me</Button>

// With variant and size
<Button variant="primary" size="md">
  Submit
</Button>

// With loading state
<Button isLoading={isSubmitting}>
  Save
</Button>

// As a link (Next.js Link)
<Button asChild variant="secondary">
  <Link href="/some-page">Go to Page</Link>
</Button>

// With icon
<Button variant="ghost" size="icon">
  <SomeIcon />
</Button>
```

## Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `variant` | `'primary' \| 'secondary' \| 'danger' \| 'ghost' \| 'link'` | `'primary'` | The visual style of the button |
| `size` | `'sm' \| 'md' \| 'lg' \| 'icon'` | `'md'` | The size of the button |
| `isLoading` | `boolean` | `false` | Whether the button is in a loading state |
| `asChild` | `boolean` | `false` | Whether to render the children directly with the button props |
| `className` | `string` | `''` | Additional CSS classes to apply |
| `disabled` | `boolean` | `false` | Whether the button is disabled |
| `...props` | `ButtonHTMLAttributes<HTMLButtonElement>` | - | All other props are passed to the underlying button element |

## Variants

### Primary

The primary button is used for the main action in a form or page. It has a solid background color and is visually prominent.

```tsx
<Button variant="primary">Primary Button</Button>
```

### Secondary

The secondary button is used for secondary actions. It has a white background with a border.

```tsx
<Button variant="secondary">Secondary Button</Button>
```

### Danger

The danger button is used for destructive actions like delete or remove. It has a red background.

```tsx
<Button variant="danger">Delete</Button>
```

### Ghost

The ghost button is used for less important actions. It has no background or border until hovered.

```tsx
<Button variant="ghost">Ghost Button</Button>
```

### Link

The link button looks like a text link but behaves like a button. It's useful for navigation-like actions within a page.

```tsx
<Button variant="link">Link Button</Button>
```

## Sizes

### Small (sm)

```tsx
<Button size="sm">Small Button</Button>
```

### Medium (md)

```tsx
<Button size="md">Medium Button</Button>
```

### Large (lg)

```tsx
<Button size="lg">Large Button</Button>
```

### Icon

```tsx
<Button size="icon">
  <SomeIcon />
</Button>
```

## Loading State

When `isLoading` is `true`, the button will display a spinner and disable user interaction.

```tsx
<Button isLoading={isSubmitting}>
  Save
</Button>
```

## Composition with Next.js Link

The `asChild` prop allows you to compose the Button with other components like Next.js Link.

```tsx
<Button asChild variant="secondary">
  <Link href="/some-page">Go to Page</Link>
</Button>
```

## Accessibility

The Button component includes proper focus states and ARIA attributes to ensure accessibility. When using the `asChild` prop with a Link component, make sure to include appropriate ARIA attributes if needed.

## Best Practices

1. Use the appropriate variant for the action:
   - `primary` for main actions
   - `secondary` for secondary actions
   - `danger` for destructive actions
   - `ghost` for less important actions
   - `link` for navigation-like actions

2. Use the appropriate size for the context:
   - `sm` for compact UIs or less important actions
   - `md` for most actions
   - `lg` for prominent actions
   - `icon` for icon-only buttons

3. Always include descriptive text or an aria-label for icon-only buttons.

4. Use the `isLoading` prop for asynchronous operations to provide visual feedback.

5. When using `asChild` with Link components, ensure the Link has appropriate attributes like `href`.
