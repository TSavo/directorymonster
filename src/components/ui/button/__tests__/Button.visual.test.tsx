import React from 'react';
import { render } from '@testing-library/react';
import { Button } from '../Button';

describe('Button Visual Regression', () => {
  it('renders all button variants', () => {
    const { container } = render(
      <div>
        <Button variant="default" data-testid="default-button">Default</Button>
        <Button variant="destructive" data-testid="destructive-button">Destructive</Button>
        <Button variant="outline" data-testid="outline-button">Outline</Button>
        <Button variant="secondary" data-testid="secondary-button">Secondary</Button>
        <Button variant="ghost" data-testid="ghost-button">Ghost</Button>
        <Button variant="link" data-testid="link-button">Link</Button>
      </div>
    );
    
    // In a real visual regression test, we would use a tool like Percy or Chromatic
    // to take a screenshot of the rendered component and compare it to a baseline
    expect(container).toMatchSnapshot();
  });

  it('renders all button sizes', () => {
    const { container } = render(
      <div>
        <Button size="default" data-testid="default-size">Default</Button>
        <Button size="sm" data-testid="sm-size">Small</Button>
        <Button size="lg" data-testid="lg-size">Large</Button>
        <Button size="icon" data-testid="icon-size">Icon</Button>
      </div>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders loading state', () => {
    const { container } = render(
      <div>
        <Button isLoading data-testid="loading-button">Loading</Button>
        <Button isLoading loadingText="Please wait..." data-testid="loading-text-button">Submit</Button>
      </div>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('renders with icons', () => {
    const { container } = render(
      <div>
        <Button 
          leftIcon={<span>🔍</span>} 
          data-testid="left-icon-button"
        >
          Search
        </Button>
        <Button 
          rightIcon={<span>→</span>} 
          data-testid="right-icon-button"
        >
          Next
        </Button>
        <Button 
          leftIcon={<span>🔍</span>} 
          rightIcon={<span>→</span>} 
          data-testid="both-icons-button"
        >
          Search and Next
        </Button>
      </div>
    );
    
    expect(container).toMatchSnapshot();
  });
});
