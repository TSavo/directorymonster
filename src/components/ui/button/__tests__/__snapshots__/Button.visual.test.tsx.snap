// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Visual Regression renders all button sizes 1`] = `
<div>
  <div>
    <button
      data-size="default"
      data-testid="default-size"
      data-variant="default"
    >
      Default
    </button>
    <button
      data-size="sm"
      data-testid="sm-size"
      data-variant="default"
    >
      Small
    </button>
    <button
      data-size="lg"
      data-testid="lg-size"
      data-variant="default"
    >
      Large
    </button>
    <button
      data-size="icon"
      data-testid="icon-size"
      data-variant="default"
    >
      Icon
    </button>
  </div>
</div>
`;

exports[`Button Visual Regression renders all button variants 1`] = `
<div>
  <div>
    <button
      data-size="default"
      data-testid="default-button"
      data-variant="default"
    >
      Default
    </button>
    <button
      data-size="default"
      data-testid="destructive-button"
      data-variant="destructive"
    >
      Destructive
    </button>
    <button
      data-size="default"
      data-testid="outline-button"
      data-variant="outline"
    >
      Outline
    </button>
    <button
      data-size="default"
      data-testid="secondary-button"
      data-variant="secondary"
    >
      Secondary
    </button>
    <button
      data-size="default"
      data-testid="ghost-button"
      data-variant="ghost"
    >
      Ghost
    </button>
    <button
      data-size="default"
      data-testid="link-button"
      data-variant="link"
    >
      Link
    </button>
  </div>
</div>
`;

exports[`Button Visual Regression renders loading state 1`] = `
<div>
  <div>
    <button
      data-loading="true"
      data-size="default"
      data-testid="loading-button"
      data-variant="default"
    >
      Loading
    </button>
    <button
      data-loading="true"
      data-size="default"
      data-testid="loading-text-button"
      data-variant="default"
    >
      Please wait...
    </button>
  </div>
</div>
`;

exports[`Button Visual Regression renders with icons 1`] = `
<div>
  <div>
    <button
      data-size="default"
      data-testid="left-icon-button"
      data-variant="default"
    >
      <span
        data-testid="left-icon"
      >
        <span>
          üîç
        </span>
      </span>
      Search
    </button>
    <button
      data-size="default"
      data-testid="right-icon-button"
      data-variant="default"
    >
      Next
      <span
        data-testid="right-icon"
      >
        <span>
          ‚Üí
        </span>
      </span>
    </button>
    <button
      data-size="default"
      data-testid="both-icons-button"
      data-variant="default"
    >
      <span
        data-testid="left-icon"
      >
        <span>
          üîç
        </span>
      </span>
      Search and Next
      <span
        data-testid="right-icon"
      >
        <span>
          ‚Üí
        </span>
      </span>
    </button>
  </div>
</div>
`;
